File: db4oj.tests/src/com/db4o/db4ounit/common/concurrency/RefreshTestCase.java
/* Copyright (C) 2004 - 2007  Versant Inc.   http://www.db4o.com */

package com.db4o.db4ounit.common.concurrency;

import com.db4o.*;
import com.db4o.ext.*;
import com.db4o.query.*;

import db4ounit.*;
import db4ounit.extensions.*;

public class RefreshTestCase extends Db4oClientServerTestCase {

	public static void main(String[] args) {
		new RefreshTestCase().runConcurrency();
	}

	public String name;

	public RefreshTestCase child;

	public RefreshTestCase() {

	}

	public RefreshTestCase(String name, RefreshTestCase child) {
		this.name = name;
		this.child = child;
	}

	protected void store() {
		RefreshTestCase r3 = new RefreshTestCase("o3", null);
		RefreshTestCase r2 = new RefreshTestCase("o2", r3);
		RefreshTestCase r1 = new RefreshTestCase("o1", r2);
		store(r1);
	}

	public void conc(ExtObjectContainer oc) {
		RefreshTestCase r11 = getRoot(oc);
		r11.name = "cc";
		oc.refresh(r11, 0);
		Assert.areEqual("cc", r11.name);
		oc.refresh(r11, 1);
		Assert.areEqual("o1", r11.name);
		r11.child.name = "cc";
		oc.refresh(r11, 1);
		Assert.areEqual("cc", r11.child.name);
		oc.refresh(r11, 2);
		Assert.areEqual("o2", r11.child.name);
	}

	private RefreshTestCase getRoot(ObjectContainer oc) {
		return getByName(oc, "o1");
	}

	private RefreshTestCase getByName(ObjectContainer oc, final String name) {
		Query q = oc.query();
		q.constrain(RefreshTestCase.class);
		q.descend("name").constrain(name);
		ObjectSet objectSet = q.execute();
		return (RefreshTestCase) objectSet.next();
	}

}

File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Concurrency/RefreshTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

#if !SILVERLIGHT
using Db4oUnit;
using Db4oUnit.Extensions;
using Db4objects.Db4o;
using Db4objects.Db4o.Ext;
using Db4objects.Db4o.Query;

namespace Db4objects.Db4o.Tests.Common.Concurrency
{
	public class RefreshTestCase : Db4oClientServerTestCase
	{
		public static void Main(string[] args)
		{
			new Db4objects.Db4o.Tests.Common.Concurrency.RefreshTestCase().RunConcurrency();
		}

		public string name;

		public Db4objects.Db4o.Tests.Common.Concurrency.RefreshTestCase child;

		public RefreshTestCase()
		{
		}

		public RefreshTestCase(string name, Db4objects.Db4o.Tests.Common.Concurrency.RefreshTestCase
			 child)
		{
			this.name = name;
			this.child = child;
		}

		protected override void Store()
		{
			Db4objects.Db4o.Tests.Common.Concurrency.RefreshTestCase r3 = new Db4objects.Db4o.Tests.Common.Concurrency.RefreshTestCase
				("o3", null);
			Db4objects.Db4o.Tests.Common.Concurrency.RefreshTestCase r2 = new Db4objects.Db4o.Tests.Common.Concurrency.RefreshTestCase
				("o2", r3);
			Db4objects.Db4o.Tests.Common.Concurrency.RefreshTestCase r1 = new Db4objects.Db4o.Tests.Common.Concurrency.RefreshTestCase
				("o1", r2);
			Store(r1);
		}

		public virtual void Conc(IExtObjectContainer oc)
		{
			Db4objects.Db4o.Tests.Common.Concurrency.RefreshTestCase r11 = GetRoot(oc);
			r11.name = "cc";
			oc.Refresh(r11, 0);
			Assert.AreEqual("cc", r11.name);
			oc.Refresh(r11, 1);
			Assert.AreEqual("o1", r11.name);
			r11.child.name = "cc";
			oc.Refresh(r11, 1);
			Assert.AreEqual("cc", r11.child.name);
			oc.Refresh(r11, 2);
			Assert.AreEqual("o2", r11.child.name);
		}

		private Db4objects.Db4o.Tests.Common.Concurrency.RefreshTestCase GetRoot(IObjectContainer
			 oc)
		{
			return GetByName(oc, "o1");
		}

		private Db4objects.Db4o.Tests.Common.Concurrency.RefreshTestCase GetByName(IObjectContainer
			 oc, string name)
		{
			IQuery q = oc.Query();
			q.Constrain(typeof(Db4objects.Db4o.Tests.Common.Concurrency.RefreshTestCase));
			q.Descend("name").Constrain(name);
			IObjectSet objectSet = q.Execute();
			return (Db4objects.Db4o.Tests.Common.Concurrency.RefreshTestCase)objectSet.Next();
		}
	}
}
#endif // !SILVERLIGHT
